File: Maths/superPalindromes.js
const isPalindrome = n => {
    let s = [...''+n];
    for (let i = 0; i < s.length/2; i++) {
        if (s[i] !== s[s.length - i - 1]) {return false}
    }
    return true
}

const superpalindromesInRange = (left, right) => {
    let lim = 100000;
    let supPs = [0];
    for (let i = 1; i < lim; i++) {
        let s = [...''+i];
        let rev = [...s].reverse()
        let odd = Number([...s, ...rev.slice(1)].join('')), even = Number([...s, ...rev].join(''))
        if (isPalindrome(odd*odd)) {supPs.push(odd*odd)}
        if (isPalindrome(even*even)) {supPs.push(even*even)} 
    }
    console.log(supPs.sort((a, b) => a - b).map(x => [x, Math.sqrt(x)]).join(', '))
}

superpalindromesInRange(0, 0)

File: Maths/superPalindromes.py
def isPalindrome(n):
    s = str(n)
    for i in range(len(s)):
        if s[i] != s[-(i+1)]:
            return False
    return True
    
def superpalindromesInRange(left, right):
    l, r = int(left), int(right)
    count = 0
    for i in range(100000):
        s = str(i)
        odd = int(s + s[-2::-1])**2
        if odd > r:
            break
        if odd >= l and isPalindrome(odd):
            count += 1
    for i in range(100000):
        s = str(i)
        even = int(s + s[::-1])**2
        if (even > r):
            break
        if even >= l and isPalindrome(even):
            count += 1
    return count

print(superpalindromesInRange(1, 100000000000000000))
        
